src/backend/commands/copyfrom.c

```diff
@@ -666,6 +682,777 @@ CopyMultiInsertInfoStore(CopyMultiInsertInfo *miinfo, ResultRelInfo *rri,
+					   TEXTOID, -1, 0);
+	TupleDescInitEntry(tupDesc, (AttrNumber) 2, "size",
+					   INT8OID, -1, 0);
+	TupleDescInitEntry(tupDesc, (AttrNumber) 3, "last_modified",
+					   TIMESTAMPTZOID, -1, 0);
+	TupleDescInitEntry(tupDesc, (AttrNumber) 4, "md5",
+					   TEXTOID, -1, 0);
+	TupleDescInitEntry(tupDesc, (AttrNumber) 5, "tag",
+					   TEXTOID, -1 ,0);
+	TupleDescInitEntry(tupDesc, (AttrNumber) 6, "file",
+					   TEXTOID, -1, 0);
+
+	num_phys_attrs = tupDesc->natts;
+
+	cstate->rel->rd_att = CreateTupleDescCopy(tupDesc);
+	cstate->rel->rd_att->tdrefcount = 1;	/* mark as refcounted */
+	cstate->attnumlist = CopyGetAttnums(tupDesc, cstate->rel, NIL);
+
+	/*
+	 * Pick up the required catalog information for each attribute in the
+	 * relation, including the input function, the element type (to pass to
+	 * the input function), and info about defaults and constraints. (Which
+	 * input function we use depends on text/binary format choice.)
+	 */
+	in_functions = (FmgrInfo *) palloc(num_phys_attrs * sizeof(FmgrInfo));
+	typioparams = (Oid *) palloc(num_phys_attrs * sizeof(Oid));
+
+	for (attnum = 1; attnum <= num_phys_attrs; attnum++)
+	{
+		Form_pg_attribute att = TupleDescAttr(tupDesc, attnum - 1);
+
+		/* We don't need info for dropped attributes */
+		if (att->attisdropped)
+			continue;
+
+		/* Fetch the input function and typioparam info */
+		getTypeInputInfo(att->atttypid,
+						 &in_func_oid, &typioparams[attnum - 1]);
+		fmgr_info(in_func_oid, &in_functions[attnum - 1]);
+	}
+
+	/* initialize progress */
+	pgstat_progress_start_command(PROGRESS_COMMAND_COPY,
+								  cstate->rel ? RelationGetRelid(cstate->rel) : InvalidOid);
+	cstate->bytes_processed = 0;
+
+	/* We keep those variables in cstate. */
+	cstate->in_functions = in_functions;
+	cstate->typioparams = typioparams;
+	cstate->is_program = false;
+
+	pipe = (filename == NULL || cstate->dispatch_mode == COPY_EXECUTOR);
+
+	if (pipe)
+	{
+		progress_vals[1] = PROGRESS_COPY_TYPE_PIPE;
+		if (whereToSendOutput == DestRemote)
+			ReceiveCopyBegin(cstate);
+		else
+			cstate->copy_file = stdin;
+	}
+	else
+	{
+		struct stat st;
+
+		cstate->filename = pstrdup(filename);
+
+		progress_vals[1] = PROGRESS_COPY_TYPE_FILE;
+		cstate->copy_file = AllocateFile(cstate->filename, PG_BINARY_R);
+		if (cstate->copy_file == NULL)
+		{
+			/* copy errno because ereport subfunctions might change it */
+			int			save_errno = errno;
+
+			ereport(ERROR,
+					(errcode_for_file_access(),
+						errmsg("could not open file \"%s\" for reading: %m",
+							   cstate->filename),
+						(save_errno == ENOENT || save_errno == EACCES) ?
+						errhint("COPY FROM instructs the PostgreSQL server process to read a file. "
+								"You may want a client-side facility such as psql's \\copy.") : 0));
+		}
+
+		// Increase buffer size to improve performance  (cmcdevitt)
+		/* GPDB_14_MERGE_FIXME: Ret value process. */
+		setvbuf(cstate->copy_file, NULL, _IOFBF, 393216); // 384 Kbytes
```

Comment:
maybe not support by cloud stroage.

Reply:
Has no problem, as in here, the copy_file should be local, which file format we only support now.

Reply:
copy local file -> local/remote directory table, copy_file is constructed from local file.